File: backend/app/src/main/java/com/elf/app/models/Employee.java
package com.elf.app.models;

import java.util.Date;
import java.util.Set;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.validator.constraints.Length;

import com.elf.app.models.utils.CivilStatus;
import com.elf.app.models.utils.EmployeeStatus;
import com.elf.app.models.utils.PublicAreaType;
import com.elf.app.models.utils.RaceType;
import com.elf.app.models.utils.SchoolingType;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Builder
@Data
@AllArgsConstructor
@NoArgsConstructor
@Table
public class Employee {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Column(unique = true, nullable = false, updatable = false)
    private String uuid;
    @Column(nullable = false, length = 350)
    private String name;
    @Column(nullable = false, length = 350)
    private String motherName;
    @Column(nullable = false, length = 350)
    private String fatherName;
    @Column(nullable = false)
    private boolean gender;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CivilStatus civilStatus;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SchoolingType schoolingType;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RaceType raceType;
    @Column(nullable = false)
    private Date birthday;
    @Column(nullable = false)
    private String nacionality;
    @Column(nullable = false)
    private String countryBirth;
    @Column(nullable = false)
    private String stateBirth;
    @Column(nullable = false)
    private String cityBirth;
    @Column(nullable = false)
    private int shoeSize;
    @Column(nullable = false)
    private int pantsSize;
    @Column(nullable = false, length = 2)
    private String shirtSize;
    @Column(nullable = false, length = 16)
    private String phoneNumber1;
    @Column(length = 16)
    private String phoneNumber2;
    @Column(nullable = false)
    private String email;
    @Column(nullable = false, length = 200)
    private String address;
    @Column(nullable = false)
    private String number;
    @Column(length = 60)
    private String complement;
    @Column(nullable = false)
    private String neighbor;
    @Column(nullable = false, length = 120)
    private String city;
    @Column(nullable = false, length = 2, columnDefinition = "char(2)")
    private String state;
    @Column(nullable = false, length = 8, columnDefinition = "char(8)")
    @Length(min = 8, max = 8)
    private String cep;
    @Column(nullable = false, length = 100)
    private String country;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PublicAreaType publicAreaType;
    @Column(nullable = false, length = 25, unique = true)
    private String rg;
    @Column(length = 10)
    private String rgIssuer;
    @Column(nullable = false, length = 80)
    private String rgIssuerState;
    @Column(nullable = false, length = 120)
    private String rgIssuerCity;
    @Column(nullable = false)
    private Date rgExpeditionDate;
    @Column(nullable = false, length = 11, unique = true)
    private String cpf;
    @Column(nullable = false, length = 11, unique = true)
    private String pis;
    @ManyToOne(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = "role_id", referencedColumnName = "id", nullable = false)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    private Role role;
    @Column(nullable = false)
    private boolean pcd;
    @Column(nullable = false)
    private boolean hosted;
    @Column(nullable = false)
    private String fileRgPath;
    @Column(nullable = false)
    private String fileCpfPath;
    @Column(nullable = false)
    private String fileCvPath;
    private String fileCnhPath;
    private String fileReservistPath;
    @Column(nullable = false)
    private boolean hasFriend;
     @Column(nullable = false)
    private String friendName;
     @Column(nullable = false)
    private String friendRole;
     @Column(nullable = false)
    private String friendCity;
    @Column(nullable = false)
    private boolean candidate;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EmployeeStatus employeeStatus;

    @OneToMany(mappedBy = "employee", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Dependent> dependents;

    @OneToMany(mappedBy = "employee", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Report> reports;

    @OneToMany(mappedBy = "employee", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<TerminationRequest> terminationRequests;

    @OneToMany(mappedBy = "employee", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<VacationRequest> vacationRequests;


    @PrePersist
    public void prePersist() {
        setUuid(java.util.UUID.randomUUID().toString());
    }
}

File: frontend/src/schemas/employee.ts
import { isCEP, isCPF, isPhone } from 'brazilian-values';
import * as z from 'zod';
import fileSchema from './file';

enum CivilStatus {
    Solteiro,
    Casado,
    Divorciado,
    Viuvo,
    Separado,
    UniãoEstável,
    Concubinato,
    Outro,
}

enum SchoolingType {
    FundamentalIncompleto,
    FundamentalCompleto,
    MédioIncompleto,
    MédioCompleto,
    SuperiorIncompleto,
    SuperiorCompleto,
    PósGraduaçãoIncompleta,
    PósGraduaçãoCompleta,
    MestradoIncompleto,
    MestradoCompleto,
    DoutoradoIncompleto,
    DoutoradoCompleto,
    PósDoutoradoIncompleto,
    PósDoutoradoCompleto,
}

enum PublicAreaType {
    Avenida,
    Rua,
    Travessa,
    Estrada,
    Rodovia,
    Alameda,
    Beco,
    Largo,
    Largueta,
    Loteamento,
    Parque,
    Praça,
    Quadra,
    Outro
}

const dependantSchema = z.object({
    cpf: z.string().min(1).refine(isCPF, {
        message: 'CPF inválido',
    }),
    gender: z.boolean(),
    name: z.string().min(1),
    birthday: z.date(),
    relationship: z.string().min(1),
})

const employeeSchema = z.object({
    name: z.string().min(1),
    motherName: z.string().min(1),
    fatherName: z.string().min(1),
    gender: z.boolean(),
    civilStatus: z.nativeEnum(CivilStatus),
    schoolingType: z.nativeEnum(SchoolingType),
    birthday: z.date(),
    nationality: z.string().min(1),
    countryBirth: z.string().min(1),
    stateBirth: z.string().min(1),
    cityBirth: z.string().min(1),
    shoeSize: z.number().min(1),
    pantsSize: z.number().min(1),
    shirtSize: z.string().min(1),
    phoneNumber1: z.string().min(1).refine(isPhone, {
        message: 'Telefone inválido',
    }),
    phoneNumber2: z.string().min(1).refine(isPhone, {
        message: 'Telefone inválido',
    }).optional(),
    email: z.string().email(),
    address: z.string().min(1),
    number: z.string().min(1),
    complement: z.string().min(1),
    neighbor: z.string().min(1),
    city: z.string().min(1),
    state: z.string().min(1),
    cep: z.string().min(1).refine(isCEP, {
        message: 'CEP inválido',
    }),
    country: z.string().min(1),
    publicAreaType: z.nativeEnum(PublicAreaType),
    rg: z.string().min(1),
    rgIssuer: z.string().min(1),
    rgIssuerState: z.string().min(1),
    rgIssuerCity: z.string().min(1),
    rgExpeditionDate: z.date(),
    cpf: z.string().min(1).refine(isCPF, {
        message: 'CPF inválido',
    }),
    pis: z.string().min(1),
    role: z.string().uuid(),
    pcd: z.boolean(),
    hosted: z.boolean(),
    fileRgPath: fileSchema,
    fileCpfPath: fileSchema,
    fileCvPath: fileSchema,
    fileCnhPath: fileSchema.optional(),
    fileReservistPath: fileSchema.optional(),
    hasFriend: z.boolean(),
    friendName: z.string().min(1).optional(),
    friendRole: z.string().min(1).optional(),
    friendCity: z.string().min(1).optional(),
    candidate: z.boolean().optional(),
    employeeStatus: z.enum([
        'Contratado',
        'Demitido',
        'Afastado',
        'Férias',
        'Aposentado',
        'Suspenso',
        'Candidato'
    ]).optional(),
    dependants: z.array(dependantSchema),
}).refine(data => {
    if (data.hasFriend) {
        return data.friendName && data.friendRole && data.friendCity;
    }
    return true;
}, "Se o funcionário tem amigo, é necessário preencher os campos de nome, cargo e cidade do amigo.")

export default employeeSchema;